drivers/gpio/gpio_sam0.c

```diff
@@ -88,23 +172,103 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
 	return 0;
 }
 
+static int gpio_sam0_manage_callback(struct device *dev,
+				     struct gpio_callback *callback, bool set)
+{
+	struct gpio_sam0_data *data = dev->driver_data;
+
+	_gpio_manage_callback(&data->cb, callback, set);
+
+	return 0;
+}
+
+static int gpio_sam0_enable_callback(struct device *dev, int access_op,
+				     u32_t pin)
+{
+	struct gpio_sam0_data *data = dev->driver_data;
+
+	if (access_op != GPIO_ACCESS_BY_PIN) {
+		return -ENOTSUP;
+	}
+
+	data->cb_pins |= 1 << pin;
+
+	return 0;
+}
+
+static int gpio_sam0_disable_callback(struct device *dev, int access_op,
+				      u32_t pin)
+{
+	struct gpio_sam0_data *data = dev->driver_data;
+
+	if (access_op != GPIO_ACCESS_BY_PIN) {
+		return -ENOTSUP;
+	}
+
+	data->cb_pins &= ~(1 << pin);
+
+	return 0;
+}
+
+static u32_t gpio_sam0_get_pending_int(struct device *dev) { return 0; }
+
+static void gpio_sam0_isr(struct device *dev)
+{
+	struct gpio_sam0_data *data = dev->driver_data;
+	u32_t flags = EIC->INTFLAG.reg;
+
+	/* Acknowledge all interrupts */
+	EIC->INTFLAG.reg = flags;
+
+	flags &= data->cb_pins;
+	if (flags != 0) {
+		_gpio_fire_callbacks(&data->cb, dev, flags);
+	}
+}
+
+int gpio_sam0_init(struct device *dev)
+{
+	/* Enable the EIC clock in PM */
+	PM->APBAMASK.bit.EIC_ = 1;
+
+	/* Enable the GCLK */
+	GCLK->CLKCTRL.reg = GCLK_CLKCTRL_ID_EIC | GCLK_CLKCTRL_GEN_GCLK0 |
+			    GCLK_CLKCTRL_CLKEN;
+
+	/* Reset the EIC */
+	EIC->CTRL.reg = EIC_CTRL_SWRST;
+	gpio_sam0_sync_eic();
+
+	/* Route and enable the interrupt.  This is safe as nothing
+	 * happens until an individual channel is unmasked.
+	 */
+	IRQ_CONNECT(EIC_IRQn, 0, gpio_sam0_isr, DEVICE_GET(gpio_sam0_0), 0);
```

Comment:
We hard-code pointer to  gpio_sam0_0 (Port A). This driver will not work correctly for gpio_sam0_1 (Port B) device. An example of using IRQ_CONNECT macro with multiple devices can be found in drivers/i2c/i2c_sam_twi.c.

Reply:
Ack.  The EIC on the SAM0 is a bit different from other chips - it's a single 16 line unit that can be shared between different ports, but that means you need some allocation / routing logic so that if both, say, PA15 and PB15 want an interrupt then it fails, and so when an EIC interrupt fires it can turn it back to a port.

This version shortcuts that by only supporting interrupts on port A.  I'll add the routing logic.

Reply:
The change is turning out to be epic :)  I've pushed a WIP version that splits the EIC out into an arch specific driver so it can be used by multiple GPIO ports.